src/diffusion/model.py

```diff
@@ -64,3 +86,80 @@ def forward(
         ).last_hidden_state
 
         return output
+
+    def training_step(self, batch, batch_idx):
+        context, gt = batch
+
+        emb = self.emb(gt)
+
+        x_0, x_t, t = get_diffusion_variables(self.diffusion_steps, emb, self.alphas_cumprod_prev)
+
+        x_0_hat = self(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_t,
+            time_ids=t,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+
+        probs = self.classifier(x_0)
+
+        ce = calculate_batch_ce(probs, gt, pad_mask)
+
+        emb_mask = ~pad_mask.unsqueeze(-1)
+
+        mse = torch.where(
+            t == 1,
+            flat_mean((x_0_hat - emb) ** 2 * emb_mask),
+            flat_mean((x_0_hat - x_0) ** 2 * emb_mask),
+        ).mean()
+
+        t0_loss = (x_0**2 * emb_mask).mean()
+
+        loss = mse + ce + t0_loss
+
+        metrics = {"mse": mse.item(), "ce": ce.item(), "t0": t0_loss.item()}
+
+        self.log("train", metrics)
+
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        context, gt = batch
+
+        emb = self.emb(gt)
+        x_0 = prepare_x0(emb)
+
+        ones = torch.ones(x_0.shape[0]).long().to(x_0.device)
+
+        for time in range(self.diffusion_steps, 1, -self.step_freq):
+            t = ones * time
+
+            x_t = get_xt(x_0, self.alphas_cumprod_prev, t)
+
+            x_0 = self(
+                encoder_input_ids=context,
+                decoder_inputs_embeds=x_t,
+                time_ids=t,
+            )
+
+        x_1 = get_xt(x_0, self.alphas_cumprod_prev, ones)
+
+        x_0 = self(
+            encoder_input_ids=context,
+            decoder_inputs_embeds=x_1,
+            time_ids=ones,
+        )
+
+        pad_mask = gt == self.emb.padding_idx
+
+        probs = self.classifier(x_0)
+
+        loss = calculate_batch_ce(probs, gt, pad_mask).item()
+
+        self.log("val_loss", loss)
```

Comment:
This way you will log eval loss on each batch. But during evaluation we often want to see average value: how good is the model on average. To do this, you need to return loss and implement `validation_epoch_end` where you will stack all outputs and calculate mean.

Reply:
And for logging to W&B, I suggest to use grouping. Passing metrics in format `val/loss`, `train/loss` will be splitter to different tabs on frontend.